File: src/main/java/com/aaronhowser1/pitchperfect/packets/SpawnNoteParticlePacket.java
package com.aaronhowser1.pitchperfect.packets;

import com.aaronhowser1.pitchperfect.client.ClientUtils;
import net.minecraft.network.FriendlyByteBuf;
import net.minecraft.resources.ResourceLocation;
import net.minecraftforge.network.NetworkEvent;

import java.util.function.Supplier;

public class SpawnNoteParticlePacket implements ModPacket{

    private final ResourceLocation soundResourceLocation;
    private final float pitch;
    private final double x;
    private final double y;
    private final double z;


    public SpawnNoteParticlePacket(ResourceLocation soundResourceLocation, float pitch, double x, double y, double z) {
        this.soundResourceLocation = soundResourceLocation;
        this.pitch = pitch;
        this.x = x;
        this.y = y;
        this.z = z;
    }

    public void encode(FriendlyByteBuf buffer) {
        buffer.writeResourceLocation(soundResourceLocation);
        buffer.writeFloat(pitch);
        buffer.writeDouble(x);
        buffer.writeDouble(y);
        buffer.writeDouble(z);
    }

    public static SpawnNoteParticlePacket decode(FriendlyByteBuf buffer) {
        return new SpawnNoteParticlePacket(
                buffer.readResourceLocation(),
                buffer.readFloat(),
                buffer.readDouble(),
                buffer.readDouble(),
                buffer.readDouble()
        );
    }

    public void receiveMessage(Supplier<NetworkEvent.Context> context) {

        ClientUtils.spawnNote(pitch, x, y, z);

        context.get().setPacketHandled(true);
    }

}

File: src/main/kotlin/com/aaronhowser1/pitchperfect/packet/SpawnNoteParticlePacket.kt
package com.aaronhowser1.pitchperfect.packet

import com.aaronhowser1.pitchperfect.utils.ClientUtils.spawnNote
import net.minecraft.network.FriendlyByteBuf
import net.minecraft.resources.ResourceLocation
import net.minecraftforge.network.NetworkEvent
import java.util.function.Supplier

class SpawnNoteParticlePacket(
    private val soundResourceLocation: ResourceLocation,
    private val pitch: Float,
    private val x: Double,
    private val y: Double,
    private val z: Double
) : ModPacket {

    override fun encode(buffer: FriendlyByteBuf) {
        buffer.writeResourceLocation(soundResourceLocation)
        buffer.writeFloat(pitch)
        buffer.writeDouble(x)
        buffer.writeDouble(y)
        buffer.writeDouble(z)
    }

    companion object {
        fun decode(buffer: FriendlyByteBuf): SpawnNoteParticlePacket {
            return SpawnNoteParticlePacket(
                buffer.readResourceLocation(),
                buffer.readFloat(),
                buffer.readDouble(),
                buffer.readDouble(),
                buffer.readDouble()
            )
        }
    }

    override fun receiveMessage(context: Supplier<NetworkEvent.Context>) {
        spawnNote(pitch, x, y, z)
        context.get().packetHandled = true
    }


}